src/vs/vscode.proposed.d.ts

```diff
@@ -1597,16 +1560,45 @@ declare module 'vscode' {
 		// first is preferred
 		// languages: string[];
 
-		// @roblourens
-		// todo@API change to `executeCells(document: NotebookDocument, cells: NotebookCellRange[], context:{isWholeNotebooke: boolean}, token: CancelationToken): void;`
-		// todo@API interrupt vs cancellation, https://github.com/microsoft/vscode/issues/106741
-		// interrupt?():void;
-		executeCell(document: NotebookDocument, cell: NotebookCell): void;
-		cancelCellExecution(document: NotebookDocument, cell: NotebookCell): void;
-		executeAllCells(document: NotebookDocument): void;
-		cancelAllCellsExecution(document: NotebookDocument): void;
+		// todo@API how can Jupyter ensure that the document-level cancel button shows whenever any cell is running?
+		// Maybe this behavior is automatic for any kernel that implements interrupt
+		interrupt?(): void;
+		executeCells(document: NotebookDocument, cells: NotebookCellRange[], cellPending: NotebookCellPending): void;
+		// executeCell(document: NotebookDocument, cell: NotebookCell): void;
+		// cancelCellExecution(document: NotebookDocument, cell: NotebookCell): void;
+		// executeAllCells(document: NotebookDocument): void;
+		// cancelAllCellsExecution(document: NotebookDocument): void;
+	}
+
+	// todo@API use the NotebookCellExecution-object as a container to model and enforce
+	// the flow of a cell execution
+
+	// kernel -> execute_info
+	// ext -> createNotebookCellExecution(cell)
+	// kernel -> done
+	// exec.dispose();
+
+	export interface NotebookCellExecution {
+		resolve(result: NotebookCellPreviousRunResult): void; // or just stop time, extension can set error outputs or not, and return a success bool
+		clearOutput(): void;
+		appendOutput(out: NotebookCellOutput): void;
+		replaceOutput(out: NotebookCellOutput): void;
+		appendOutputItems(output: string, items: NotebookCellOutputItem[]): void;
+		replaceOutputItems(output: string, items: NotebookCellOutputItem[]): void;
+		token: CancellationToken;
 	}
 
+	export interface NotebookCellPending {
```

Comment:
It makes things very explicit but I am unsure if we are over-engineering. Should `NotebookCellExecution#resolve` then return a notebook cell end state? 

An alternative would be to incorporate this into the cell execution, e.g `start` on there and manage the state internally

Reply:
That would probably be ok, as long as we have only two states here and you can't do anything special in the pending state, it's not too confusing.

Reply:
> Should NotebookCellExecution#resolve then return a notebook cell end state?

It could but there's no point to it